### DISCUSSION TITLE
Multiprocessing of RandCropByPosNegLabel

### DISCUSSION THREAD

**Discussion by StefanFischer** [Project-MONAI/MONAI#7964] [Q&A]
Hello MONAI-Team and Users,

i would like to know if there is an option to run RandCropByPosNegLabel Transforms directly on the GPU with multiprocessing.
To ensure fast dataloading i use MONAI's CacheDataset to do determistic transforms before training start. During training the RandCropByPosNegLabel transform is applied on the GPU. I try to train with a very high batch size (>1000 samples per batch) and then the dataloading gets quite slow (5s per iteration) and minimal patch size (16x16x16, bs=1028). While the number of voxels per training input tensor is equal to the maximal patch size (128x128x128, bs=2), the training with the small patch size is significantly larger than with the maximal patch size. Besides the RandCropByPosNegLabel transformation i do NOT apply any other random transform.

Is there an option to do it more efficiently than currently implemented in the MONAI library?
 The code of the RandCropByPosNegLabel shows a for loop over the cropping operation. 
https://github.com/Project-MONAI/MONAI/blob/59a7211070538586369afd4a01eca0a7fe2e742e/monai/transforms/croppad/array.py#L1047-L1223. I think that it is possible to parralize the following code:
`for i, center in enumerate(self.centers):
                cropper = SpatialCrop(roi_center=center, roi_size=roi_size, lazy=lazy_)
                cropped = cropper(img)
`

I tried to use multiprocessing (torch.multiprocessing), but i always get CUDA related issues. I am also an absolute noob in multiprocessing. Does anyone has experience in CUDA multiprocessing?
Would be happy to hear from you guys :)

**Comment by KumoLiu** [comment]
Hi @StefanFischer, to put transforms on the GPU, you can try `ThreadDataLoader `.

For more details, you can refer to this tutorial: https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_model_training_guide.md

Hope it helps, thanks.

**Reply by StefanFischer** [reply]
Hi KumoLiu,

thanks for the fast reply!
I already use the `ThreadDataLoader`, but i thought that the whole process of the RandCropByPosNegLabel transform can be implemented faster using multiprocessing.
Instead of iterating through the crop creation in a serial fashion (for-loop) as in the current MONAI implementation (class RandCropByPosNegLabel)
`
for i, center in enumerate(self.centers): 
    cropper = SpatialCrop(roi_center=center, roi_size=roi_size, lazy=lazy_) 
    cropped = cropper(img)
`
would it be also possible to do crop creation using multiprocessing (in parallel)?
I only got CUDA multiprocessing errors when I tried to parallelize it. I guess there are issues with Memory Managment on GPU tensors.

Thank you for the help!

Best,
Stefan

**Reply by KumoLiu** [reply]
Hi @StefanFischer, 

Yes, using multiprocessing for image cropping tasks can significantly improve performance, especially when dealing with large datasets or computationally intensive operations. However, as you mentioned, working with GPU tensors and CUDA in a multiprocessing context can be tricky due to memory management issues. To avoid these issues, one approach is to perform the cropping operations on the CPU using multiprocessing and then transfer the cropped tensors to the GPU if needed. But this way it may increase the time of the IO operations.

You can also refer to the discussion here: https://github.com/Project-MONAI/MONAI/issues/2794
We didn't enable multi-threads for ThreadDataloader because most of MONAI random transforms are not thread-safe and definitely need to enhance our logic.

Thanks.
